### PR TITLE
Add websub sink and source implementations

### PR DESCRIPTION
**Pull request by AnuGayan**
## Purpose
 This PR will add webSub sink and source implementation. 

### Source
Using webSubHub Source, Users can subscribe to the topics, which are defined under the webSub hub. Each webSubHub must have a unique hub.id`, `table.name` .  
WebHubSub Source MUST be followed by a` @store` annotation which configured to an external database. This Database will be used to create the subscription database and store the subscription related data.
The same table can be shared through several Hubs but a unique table is preferred. 
Following parameter MUST be there in the source stream and will be mapped the incoming request to the given stream parameters

> callback string - callback URL
> lease_seconds long -  subscription lease seconds
> secret string  - the cryptographically random unique secret string that will be used to compute an HMAC digest for authorized content distribution
> topic string - the topic URL that the subscriber wishes to subscribe to or unsubscribe from
> mode string - The literal string "subscribe" or "unsubscribe", depending on the goal of the request.

Subscription request must follow the [WebSub specifications](https://www.w3.org/TR/websub/#x5-1-subscriber-sends-subscription-request) and will be rejected if not 

### Sink

WebSubHubEventPublisher sink MUST be followed by a` @store` annotation which configured to a external datasource which subscription table resides
WebSubHubEventPublisher MUST contains a `hub.id` and a` table.name`. And always the `keyvalue `map has to be used.
WebSubHubEventPublisher stream MUST have the `topic string,` `payload string` parameters and if there are any other parameters, they will be ignored

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
 WebSubSink 

```
@store(type='rdbms' , jdbc.url='jdbc:mysql://localhost:3306/production?useSSL=false', username='root', password='root', jdbc.driver.name='com.mysql.jdbc.Driver') 
@source(type='WebSubHub' , hub.id='anu_123',table.name='SessionTable', receiver.url='http://localhost:8006/productionStream',basic.auth.enabled='false', @map(type='keyvalue',implicit.cast.enable='true', @payload('payload_string'))) 
define stream webSubStream(callback string, lease_seconds long, secret string, topic string, mode string);
```

WebSubHubEventPublisher

```
@store(type='rdbms' , jdbc.url='jdbc:mysql://localhost:3306/production?useSSL=false', username='root', password='root', jdbc.driver.name='com.mysql.jdbc.Driver') 
@sink(type='WebSubHubEventPublisher', hub.id="anu_123" , table.name='SessionTable'
@map(type='keyvalue'))
define stream LowProductionAlertStream (topic string, payload string);
```

### CODE REVIEW

File: component/src/main/java/io/siddhi/extension/io/http/sink/WebSubHubSink.java [hunk 1]
```diff
@@ -0,0 +1,904 @@
...
+                if (System.currentTimeMillis() < subscribedTime + subscriptionLeaseTime) {
+                    if (!webSubDTO.getSecret().equalsIgnoreCase("")) {
+                        String signatureHeaderValue = SHA256_HASHING +
+                                getHexValue(payloadMap.get("payload").toString());
+                        Header signatureHeader = new Header(X_HUB_SIGNATURE, signatureHeaderValue);
+                        headersList.add(signatureHeader);
+                    }
+
+                    String publisherURL = webSubDTO.getCallback();
+
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/http/sink/WebSubHubSink.java:525]: Remove new lines

File: component/src/main/java/io/siddhi/extension/io/http/sink/WebSubHubSink.java [hunk 2]
```diff
@@ -0,0 +1,904 @@
...
+        BufferedInputStream bis = new BufferedInputStream(in);
+        String result;
+        try (ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
+            int data;
+            while ((data = bis.read()) != -1) {
+                bos.write(data);
+            }
+            result = bos.toString(StandardCharsets.UTF_8.toString());
+        } catch (IOException ioe) {
+            log.error("Couldn't read the complete input stream");
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/http/sink/WebSubHubSink.java:580]: add the exception and error trance to the log

File: component/src/main/java/io/siddhi/extension/io/http/sink/WebSubHubSink.java [hunk 3]
```diff
@@ -0,0 +1,904 @@
...
+                        List<WebSubSubscriptionDTO> webSubSubscriptionList = tempWebSubSubscriptionDTOMap.
+                                getOrDefault(topic, new ArrayList<>());
+                        webSubSubscriptionList.add(new WebSubSubscriptionDTO(hubId, event.getData(0),
+                                topic, event.getData(2), event.getData(3), event.getData(4)));
+                        tempWebSubSubscriptionDTOMap.put(topic, webSubSubscriptionList);
+                    }
+                }
+                setWebSubSubscriptionMap(tempWebSubSubscriptionDTOMap);
+
+                log.info(Arrays.toString(events));
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/http/sink/WebSubHubSink.java:898]: do we need this ?

File: component/src/main/java/io/siddhi/extension/io/http/source/HttpConnectorRegistry.java [hunk 4]
```diff
@@ -143,6 +145,23 @@ void registerSourceListener(SourceEventListener sourceEventListener, String list
...
+                                SiddhiAppContext siddhiAppContext) {
+        String listenerKey = HttpSourceUtil.getSourceListenerKey(listenerUrl, metrics);
+        HttpSourceListener httpSourceListener = this.sourceListenersMap.putIfAbsent(listenerKey,
+                new HttpSourceListener(workerThread, listenerUrl, isAuth, sourceEventListener,
+                        requestedTransportPropertyNames, siddhiAppName, metrics, table, hubId, siddhiAppContext));
+        if (httpSourceListener != null) {
+            if (metrics != null) {
+                metrics.getTotalHttpErrorsMetric().inc();
+            }
+
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/http/source/HttpConnectorRegistry.java:160]: remove new line

File: component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubResponseProcessor.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            logger.error("Error occurred while processing the payload ", e);
+        } finally {
+            try {
+                buf.close();
+                carbonMessage.waitAndReleaseAllEntities();
+            } catch (IOException e) {
+                if (metrics != null) {
+                    metrics.getTotalHttpErrorsMetric().inc();
+                }
+
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubResponseProcessor.java:197]: remove new line

File: component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubResponseProcessor.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        MetaStateEvent tableUpdateMetaStateEvent = new MetaStateEvent(2);
+
+        MetaStreamEvent tableMetaStreamEvent = new MetaStreamEvent();
+        MetaStreamEvent inputStreamMetaStreamEvent = new MetaStreamEvent();
+
+        TableDefinition inputTableDefinition = TableDefinition.id("");
+        table.getTableDefinition().getAttributeList().forEach((
+                attribute -> inputTableDefinition.attribute(attribute.getName(), attribute.getType())));
+        inputStreamMetaStreamEvent.addInputDefinition(inputTableDefinition);
+
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubResponseProcessor.java:236]: remove

File: component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubSource.java [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+import static io.siddhi.extension.io.http.util.HttpConstants.DEFAULT_WORKER_COUNT;
+import static io.siddhi.extension.io.http.util.HttpConstants.HUB_CALLBACK;
+import static io.siddhi.extension.io.http.util.HttpConstants.HUB_ID;
+import static io.siddhi.extension.io.http.util.HttpConstants.HUB_ID_COLUMN_NAME;
+import static io.siddhi.extension.io.http.util.HttpConstants.HUB_TOPIC;
+import static io.siddhi.extension.io.http.util.HttpConstants.REQUEST_TIMESTAMP;
+import static io.siddhi.extension.io.http.util.HttpConstants.SOCKET_IDEAL_TIMEOUT_VALUE;
+import static io.siddhi.extension.io.http.util.HttpConstants.WEB_SUB_SUBSCRIPTION_DATA_TABLE_DEFAULT_NAME;
+
+/**
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubSource.java:63]: add a description

**Review by dnwick — CHANGES_REQUESTED** [review]

File: component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubResponseProcessor.java (hunk 6 above)
Comment (AnuGayan) [reply component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubResponseProcessor.java:234]: I have added this new line to separate the two tasks for better readability.

File: component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubSource.java (hunk 7 above)
Comment (AnuGayan) [reply component/src/main/java/io/siddhi/extension/io/http/source/HttpWebSubSource.java:63]: Done

**Review by dnwick — APPROVED** [review]

**Pull request merged**